src/coreclr/jit/lsraarmarch.cpp

```diff
@@ -368,6 +368,21 @@ int LinearScan::BuildCall(GenTreeCall* call)
 
     if (ctrlExpr != nullptr)
     {
+#ifdef TARGET_ARM64
+        if (compiler->IsTargetAbi(CORINFO_NATIVEAOT_ABI) && TargetOS::IsUnix && (call->gtArgs.CountArgs() == 0) &&
+            ctrlExpr->IsTlsIconHandle())
+        {
+            // For NativeAOT linux/arm64, we generate the needed code as part of
+            // call node because the generated code has to be in specific format
+            // that linker can patch. As such, the code needs specific registers
+            // that we will attach to this node to guarantee that they are available
+            // during generating this node.
+            assert(call->gtFlags & GTF_TLS_GET_ADDR);
+            newRefPosition(REG_R0, currentLoc, RefTypeFixedReg, nullptr, genRegMask(REG_R0));
```

Comment:
Do these have to be R0/R1/R2 for the linker, or is this just a choice you've made to make things simpler?

Or is it the case that since the ultimate expansion has a call these choices don't add new constraints?

Reply:
They are used by the linker to match the pattern as seen [here](https://github.com/llvm/llvm-project/blob/fb48fd18c240574841378defacadff34238089bb/lld/ELF/Arch/AArch64.cpp#L539-L551) and thats how we did that in hand-assembly code [here](https://github.com/dotnet/runtime/pull/97910/files#diff-5ba96261757fec78bede1a85221149817692173fa2c24a7357ddd213751290b1L263-L283).